packages/protocol/contracts/ACE/validators/publicRange/PublicRange.sol

```diff
@@ -0,0 +1,351 @@
+
+                    if gt(i, 0) {
+                        // Set k = kx_j, a = ax_j, c = cx_j, where j = i - (m+1)
+                        let x := mod(mload(0x00), gen_order) // x is the kecca hash of the input commitments
+                        k := mulmod(k, x, gen_order) // kx
+                        a := mulmod(a, x, gen_order) // ax
+                        c := mulmod(challenge, x, gen_order) // cx
+                    }
+                    
+                    // Calculate the G1 element \gamma_i^{k}h^{a}\sigma_i^{-c} = B_i
+                    // Memory map:
+                    // 0x20: \gamma_iX
+                    // 0x40: \gamma_iY
+                    // 0x60: k_i
+                    // 0x80: hX
+                    // 0xa0: hY
+                    // 0xc0: a_i
+                    // 0xe0: \sigma_iX
+                    // 0x100: \sigma_iY
+                    // 0x120: -c
+
+
+                    // * Note data map (uint[6]) is
+                    // * 0x00:0x20       = Z_p element \bar{k}_i
+                    // * 0x20:0x40       = Z_p element \bar{a}_i
+                    // * 0x40:0x80       = G1 element \gamma_i
+                    // * 0x80:0xc0       = G1 element \sigma_i
+
+                    // loading into memory
+                    calldatacopy(0xe0, add(noteIndex, 0x80), 0x40)
+                    calldatacopy(0x20, add(noteIndex, 0x40), 0x40)
+                    mstore(0x120, sub(gen_order, c)) 
+                    mstore(0x60, k)
+                    mstore(0xc0, a)
+
+                    // Call bn128 scalar multiplication precompiles
+                    // Represent point + multiplication scalar in 3 consecutive blocks of memory
+                    // Store \sigma_i^{-c} at 0x1a0:0x200
+                    // Store \gamma_i^{k} at 0x120:0x160
+                    // Store h^{a} at 0x160:0x1a0
+                    let result := staticcall(gas, 7, 0xe0, 0x60, 0x1a0, 0x40) // sigmai^-c
+                    result := and(result, staticcall(gas, 7, 0x20, 0x60, 0x120, 0x40))
+                    result := and(result, staticcall(gas, 7, 0x80, 0x60, 0x160, 0x40))
+
+                    // Call bn128 group addition precompiles
+                    // \gamma_i^{k} and h^{a} in memory block 0x120:0x1a0
+                    // Store result of addition at 0x160:0x1a0
+                    result := and(result, staticcall(gas, 6, 0x120, 0x80, 0x160, 0x40))
+
+                    // \gamma_i^{k}h^{a} and \sigma^{-c} in memory block 0x160:0x1e0
+                    // Store resulting point B at memory index b
+                    result := and(result, staticcall(gas, 6, 0x160, 0x80, b, 0x40))
+
+                    // If i > 0 (i.e. if it is the utility note)
+                    // then we add \sigma^{-c} and \sigma_{acc} and store result at \sigma_{acc} (0x1e0:0x200)
+                    // we then calculate \gamma^{cx} and add into \gamma_{acc}
+                    if gt(i, 0) { // m = 0
```

Comment:
it doesn't look like we're checking the range of the input note. We should probably do so, we can't assume that the input note has come from an AZTEC note registry

Reply:
yeah good point. I have now removed the `if` statement and I roll the `gamma` and `sigma` coordinates of all notes into the accumulators. 

The pairing check is then called on the accumulators